### PR TITLE
feat: route OpenAI Responses and Anthropic providers through GitHub Copilot

### LINKED ISSUE #63: Route OpenAI Responses & Anthropic providers through GitHub Copilot
## Summary
Add a shared GitHub Copilot transport so the existing OpenAI Responses and Anthropic providers can use a GitHub Copilot subscription as their backend (`api.enterprise.githubcopilot.com`), and fix the Responses-API tool-calling path (broken end-to-end) that this surfaced.

## Motivation
Reuse an existing GitHub Copilot login (no separate OpenAI/Anthropic API keys) for both `/v1/messages` (Anthropic) and `/responses` (OpenAI), including the WebSocket transport the Copilot CLI uses.

## Scope
- Shared transport in LmCore: `ICopilotTokenProvider` (CLI credential reader + device-flow fallback), `CopilotHeadersHandler`, `CopilotSessionContext`, `CopilotOptions`, `CopilotHttpClientFactory`.
- Anthropic `/v1/messages` via Copilot (`CopilotAnthropicAgentFactory`).
- OpenAI `/responses` via Copilot over SSE and WebSocket (`CopilotResponsesAgentFactory`, `CopilotResponsesWebSocketClient`).
- LmStreaming.Sample: four new selectable providers (Sonnet, Haiku, GPT-5.5, GPT-5.5 mini).
- Fixes surfaced by real Copilot traffic: `tools[].parameters` schema object, multi-turn tool round-trip (Composite/Aggregate unpacking), `output_item.done` tool emission, Anthropic out-of-order `type` discriminator.

## Acceptance
- Both providers answer live through Copilot.
- Tool calls complete without looping.
- Unit + live tests pass.

### PR DESCRIPTION
**Pull request by gautam-achieveai**
## Summary
Adds a shared GitHub Copilot transport so the existing **OpenAI Responses** and **Anthropic** providers can use a GitHub Copilot subscription as their backend (`api.enterprise.githubcopilot.com`), reusing a Copilot login instead of separate OpenAI/Anthropic keys. Along the way it fixes the Responses-API tool-calling path, which was broken end-to-end (a 400 on `tools[].parameters`, then an infinite tool-call loop).

The protocol was reverse-engineered from captured GitHub Copilot CLI traffic; the raw `gho_` token is used directly (no token exchange).

## Changes
**Shared transport (LmCore)**
- `ICopilotTokenProvider` with a CLI-credential reader (env vars, Copilot CLI `config.json` (JSONC), `gh` `hosts.yml`, `gh auth token`) and a device-flow fallback.
- `CopilotHeadersHandler` (bearer auth + `copilot-integration-id`/`editor-version`/`x-github-api-version`/`user-agent`, fresh `x-interaction-id` per request, stable machine/session ids), `CopilotSessionContext`, `CopilotOptions`, `CopilotRequestHeaders`, `CopilotHttpClientFactory`.

**Providers**
- Anthropic `/v1/messages` via Copilot — `CopilotAnthropicAgentFactory` (reuses `AnthropicAgent` unchanged).
- OpenAI `/responses` via Copilot over **SSE and WebSocket** — `CopilotResponsesAgentFactory`, `CopilotResponsesWebSocketClient` (multi-turn `previous_response_id` chaining); configurable responses path; new `reasoning`/`text`/`store`/`parallel_tool_calls`/`initiator` request fields.
- `LmStreaming.Sample`: four new selectable providers — Sonnet, Haiku, GPT-5.5, GPT-5.5 mini.

**Bug fixes (pre-existing, surfaced by real Copilot traffic)**
- `tools[].parameters` is now a JSON Schema object, not an array (fixes a 400).
- Tool round-trip: `function_call` carries `name`+`arguments`; `MessageMapper` recurses `CompositeMessage` and unpacks `ToolsCallAggregateMessage` so the replayed tool result actually reaches the model (fixes an infinite tool-call loop).
- Agent emits tool calls from `output_item.done` (Copilot rotates the per-event `item_id`), deduped by `call_id`.
- Anthropic deserializer tolerates an out-of-order `type` discriminator (net9 `AllowOutOfOrderMetadataProperties`).

## Testing
- Unit/integration: full solution green — incl. `OpenAiResponsesProvider.Tests`, `AnthropicProvider.Tests` (166), `LmCore.Tests` (741, incl. new Copilot auth tests). New tests cover token providers, header injection, mapper tool round-trip (composite/aggregate), and WebSocket framing/chaining. (~2,460 passing, 0 real failures.)
- E2E: `LmStreaming.Sample.E2E.Tests` 10/10; `LmStreaming.Sample.Browser.E2E.Tests` 13 passed / 4 skipped (CLI-gated).
- Live (manual `CopilotLive.Tests`, excluded from CI, gated on a Copilot login): 6/6 against the real API — Anthropic stream + non-stream, Responses SSE + WebSocket, multi-turn chaining.
- Verified in the sample app: Sonnet/Haiku/GPT-5.5/GPT-5.5 mini answer; calculator tool round-trip completes ("17 plus 25 equals 42.") with no loop.

Note: `CopilotLive.Tests` is intentionally **not** in the solution / CI; run it on demand with `dotnet test tests/CopilotLive.Tests`. Defaults mimic the Copilot CLI's integration headers (a ToS consideration).

## Related Issues
Fixes #63

### CODE REVIEW

**Review by gautam-achieveai — COMMENTED** [review]
## [Gautam's review bot] Deep Review — PR #64

**Work Item:** #63 — Route OpenAI Responses & Anthropic providers through GitHub Copilot

**Build:** ✅ Passes (after NuGet restore)

**Agents used:** Architecture Review, Performance Review, Exception Handling Review, Test Coverage Review

---

### Summary

Well-structured PR with clean composition patterns — the factories compose existing components (AnthropicClient, OpenAiResponsesClient) with only transport configuration changes. The `ICopilotResponsesSocket` abstraction, shared `CopilotRequestHeaders.Build()`, and DelegatingHandler pipeline are all solid design choices.

However, there are **blocking concerns** around thread safety in disposal, missing test coverage for the PR's core behavioral additions, and a correctness gap in exception propagation.

---

### Findings

| Sev | Finding | File |
|-----|---------|------|
| 🔴 HIGH | Sync-over-async `Dispose()` — network I/O with no timeout on blocked thread | `CopilotResponsesWebSocketClient.cs:188` |
| 🔴 HIGH | `CompositeCopilotTokenProvider` has **zero** unit tests (concurrency, fallback, caching) | `CompositeCopilotTokenProvider.cs` |
| 🔴 HIGH | `output_item.done` dedup path (the Copilot-specific fix) has no isolated test | `OpenAiResponsesAgent.cs:270-303` |
| 🟡 MEDIUM | Token cached forever — no TTL/refresh for long-lived processes | `CompositeCopilotTokenProvider.cs` |
| 🟡 MEDIUM | Silent WebSocket exception → `null` (treated as graceful close, no logging) | `ClientWebSocketResponsesSocket.cs:261` |
| 🟡 MEDIUM | Composite provider loses all exceptions except the last | `CompositeCopilotTokenProvider.cs:68` |
| 🟡 MEDIUM | `CopilotHttpClientFactory` hard-codes `new HttpClientHandler()` (can't inject for tests/Polly) | `CopilotHttpClientFactory.cs:33` |
| 🟡 MEDIUM | Per-call 16KB buffer allocation in WebSocket receive (use `ArrayPool`) | `ClientWebSocketResponsesSocket` |
| 🟡 MEDIUM | `ProviderRegistry` calls `ResolveToken()` synchronously — potential 5s first-request delay | `ProviderRegistry.cs:73` |
| 🟢 LOW | Copilot Auth subsystem (~600 LOC) in LmCore widens core with vendor-specific code | `src/LmCore/Auth/` |
| 🟢 LOW | `ResolveCacheFilePath` dead catch (IOException can't occur from Path.Combine) | `DeviceFlowCopilotTokenProvider.cs:197` |

---

### Minimum for Approval

1. **Fix the Dispose sync-over-async** — either implement `IAsyncDisposable` or add a timeout guard (`task.Wait(2s)`) + ensure `_turnGate.Dispose()` runs in a finally block.
2. **Add `CompositeCopilotTokenProvider` unit tests** — at minimum: first-wins precedence, fallback on failure, in-memory caching, cancellation propagation.
3. **Add a targeted test for the `output_item.done`-only path** — verify that when deltas have rotated item_ids (the Copilot scenario), the `OutputItemDone` event alone produces a `ToolsCallMessage`.

---

### Positive Observations

- Clean dependency direction — providers depend on LmCore inward, no reverse deps
- Excellent `ICopilotResponsesSocket` testability abstraction with `FakeSocket` in tests
- `CopilotRequestHeaders.Build()` shared between HTTP handler and WebSocket — prevents drift
- `MessageMapper` refactoring to handle `CompositeMessage`/`ToolsCallAggregateMessage` recursion solves the infinite tool-call loop correctly
- `BuildParametersSchema()` properly converts the parameter list to a JSON Schema object
- Live test suite with `[SkippableFact]` is a good pattern for manual smoke testing

File: src/OpenAiResponsesProvider/Agents/CopilotResponsesWebSocketClient.cs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    /// <inheritdoc />
+    public void Dispose()
+    {
+        if (_disposed)
+        {
+            return;
+        }
+
+        _disposed = true;
+        _socket?.DisposeAsync().AsTask().GetAwaiter().GetResult();
```
Comment (gautam-achieveai) [inline src/OpenAiResponsesProvider/Agents/CopilotResponsesWebSocketClient.cs]: **[Gautam's review bot] HIGH** - Sync-over-async Dispose blocks thread with no timeout. `.DisposeAsync().AsTask().GetAwaiter().GetResult()` performs a network round-trip with CancellationToken.None on a blocked thread. Also, if DisposeAsync() throws, _turnGate.Dispose() is never reached. Fix: wrap in try/finally with bounded Wait(2s), or implement IAsyncDisposable.

File: src/LmCore/Auth/CompositeCopilotTokenProvider.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                "No GitHub Copilot token could be resolved from any configured provider.",
+                last
+            );
+        }
+        finally
+        {
+            _ = _gate.Release();
+        }
+    }
+}
```
Comment (gautam-achieveai) [inline src/LmCore/Auth/CompositeCopilotTokenProvider.cs:85]: **[Gautam's review bot] HIGH** - This class has zero unit tests. It's the primary token resolution entry point with thread-safe double-checked locking, first-wins fallback, caching, and exception handling. Please add tests for: first-wins precedence, fallback on failure, in-memory caching, and cancellation propagation.

File: src/LmCore/Auth/CompositeCopilotTokenProvider.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                {
+                    var token = await provider.GetTokenAsync(cancellationToken).ConfigureAwait(false);
+                    if (!string.IsNullOrWhiteSpace(token))
+                    {
+                        _cachedToken = token;
+                        return token;
+                    }
+                }
+                catch (Exception ex) when (ex is not OperationCanceledException)
+                {
```
Comment (gautam-achieveai) [inline src/LmCore/Auth/CompositeCopilotTokenProvider.cs:68]: **[Gautam's review bot] MEDIUM** - Only the last exception is preserved when all providers fail. Earlier failures (often more informative) are lost. Consider accumulating into a List<Exception> and wrapping in AggregateException.

File: src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs [hunk 4]
```diff
@@ -261,6 +267,42 @@ [EnumeratorCancellation] CancellationToken cancellationToken
```
Comment (gautam-achieveai) [inline src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs:267]: **[Gautam's review bot] HIGH** - This output_item.done fallback path (the core Copilot-specific fix) lacks a targeted unit test. Please add a test that yields events where delta item_ids DON'T match (simulating Copilot), then an output_item.done with complete function_call data, and asserts exactly one ToolsCallMessage is produced.

File: src/LmCore/Auth/CopilotHttpClientFactory.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        CopilotSessionContext session,
+        CopilotOptions? options = null,
+        TimeSpan? timeout = null
+    )
+    {
+        ArgumentNullException.ThrowIfNull(baseAddress);
+        ArgumentNullException.ThrowIfNull(tokenProvider);
+        ArgumentNullException.ThrowIfNull(session);
+
+        var handler = new CopilotHeadersHandler(tokenProvider, session, options, new HttpClientHandler());
```
Comment (gautam-achieveai) [inline src/LmCore/Auth/CopilotHttpClientFactory.cs]: **[Gautam's review bot] MEDIUM** - Hard-coded `new HttpClientHandler()` blocks testability/extensibility. Add an optional `HttpMessageHandler? innerHandler = null` parameter so callers can inject test mocks, Polly resilience, or IHttpClientFactory integration.

File: src/OpenAiResponsesProvider/Agents/CopilotResponsesWebSocketClient.cs (hunk 1 above)
Comment (gautam-achieveai) [reply src/OpenAiResponsesProvider/Agents/CopilotResponsesWebSocketClient.cs]: [Gautam's bot] Resolved in 2bb2a2a — `CopilotResponsesWebSocketClient` now implements `IAsyncDisposable`, and the synchronous `Dispose()` performs a bounded `Wait(2s)` inside a try/finally so the socket close can't hang the thread and `_turnGate.Dispose()` always runs.

File: src/LmCore/Auth/CompositeCopilotTokenProvider.cs (hunk 2 above)
Comment (gautam-achieveai) [reply src/LmCore/Auth/CompositeCopilotTokenProvider.cs:85]: [Gautam's bot] Resolved in 2bb2a2a — added `CompositeCopilotTokenProviderTests` covering first-wins precedence, fallback on failure, in-memory caching (provider invoked once), cancellation propagation (OCE not swallowed), and all-fail preserving every error.

File: src/LmCore/Auth/CompositeCopilotTokenProvider.cs (hunk 3 above)
Comment (gautam-achieveai) [reply src/LmCore/Auth/CompositeCopilotTokenProvider.cs:68]: [Gautam's bot] Resolved in 2bb2a2a — failures now accumulate into a `List<Exception>` wrapped in an `AggregateException` as the `InnerException`, so earlier (often more informative) errors are preserved. Covered by the new all-fail test.

File: src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs (hunk 4 above)
Comment (gautam-achieveai) [reply src/OpenAiResponsesProvider/Agents/OpenAiResponsesAgent.cs:267]: [Gautam's bot] Resolved in 2bb2a2a — added `OpenAiResponsesAgentToolCallTests`: one test drives mismatched/rotated delta `item_id`s (the Copilot case) and asserts the `output_item.done` path yields exactly one `ToolsCallMessage`; a second asserts the delta + output_item.done paths dedupe by `call_id`.

File: src/LmCore/Auth/CopilotHttpClientFactory.cs (hunk 5 above)
Comment (gautam-achieveai) [reply src/LmCore/Auth/CopilotHttpClientFactory.cs]: [Gautam's bot] Resolved in 2bb2a2a — `CopilotHttpClientFactory.Create` now takes an optional `HttpMessageHandler? innerHandler = null` (defaults to `new HttpClientHandler()`) for test mocks / Polly / IHttpClientFactory.

**Comment by gautam-achieveai** [comment]
[Gautam's bot] Addressed the deep-review feedback in `2bb2a2a` (verified building + testing under .NET SDK 9, CI's SDK).

**Blockers — resolved (minimum for approval):**
1. ✅ WS `Dispose()` sync-over-async → `IAsyncDisposable` + bounded 2s `Wait` in try/finally.
2. ✅ `CompositeCopilotTokenProvider` unit tests (5) — precedence, fallback, caching, cancellation, all-fail.
3. ✅ `output_item.done` tool-call path tests (2) — rotated `item_id`s yield one call; delta+done dedupe.

**MEDIUM — resolved:**
- ✅ Composite now aggregates all provider failures (`AggregateException`).
- ✅ `CopilotHttpClientFactory` accepts an injectable `HttpMessageHandler`.

**Deferred (with rationale — happy to file follow-ups if you'd prefer):**
- 🟡 *Token TTL/refresh* — by design this PR sends the raw `gho_` GitHub OAuth token directly (no exchange), which is long-lived; a refresh hook belongs with the future `/copilot_internal/v2/token` exchange path (individual host). YAGNI today.
- 🟡 *ProviderRegistry sync `ResolveToken`* — runs once at sample-app startup reading CLI creds (fast); the 5s ceiling is only the worst-case `gh auth token` shell-out and never on the request hot path. Sample-only.
- 🟡 *ArrayPool for the 16KB WS receive buffer* — one allocation per turn (not hot); micro-opt deferred.
- 🟡 *Silent WS exception → null* — the socket abstraction has no logger; surfaced at the client layer. Can add a trace log if desired.
- 🟢 *LmCore `Auth/` placement* — deliberate so both providers share the transport without a new package; extractable later.
- 🟢 *`ResolveCacheFilePath` defensive catch* — `GetFolderPath` can throw on locked-down environments; left as a guard.

CI re-running on the new commit.

**Comment by gautam-achieveai** [comment]
**[Gautam's review bot]** Re-review of iteration 3 (`343e86fa` → `2bb2a2ab`).

All **3 HIGH blockers** and **2 MEDIUM items** from the initial deep review are resolved:

| Finding | Status |
|---------|--------|
| Sync-over-async `Dispose()` (thread starvation) | ✅ `IAsyncDisposable` + bounded 2s `Wait` in try/finally |
| `CompositeCopilotTokenProvider` zero tests | ✅ 5 focused tests (precedence, fallback, cache, cancel, all-fail) |
| `output_item.done` tool-call path untested | ✅ 2 tests (rotated item_ids, delta+done dedup) |
| Composite loses exceptions except last | ✅ `AggregateException` wraps all failures |
| `CopilotHttpClientFactory` hard-coded handler | ✅ Optional `innerHandler` parameter |

**Fix quality notes:**
- `DisposeAsync` correctly disposes socket then turn gate via try/finally — clean separation of the async-preferred and sync-bounded paths.
- Tests are well-structured with minimal stubs (`ScriptedClient`, `StubProvider`) and precise assertions.
- `AggregateException` approach properly preserves the full failure chain for diagnostics.

**Deferred items** (token TTL, ArrayPool, silent WS exception, Auth/ placement, sync `ResolveToken` in sample) have reasonable justification and are acceptable as follow-ups.

Build: ✅ 0 warnings, 0 errors.

**Verdict: APPROVE** — All blockers resolved. Good work addressing the feedback.
